### PR TITLE
Preserve OrderedDict outputs in FX tracing

### LINKED ISSUE #161807: torch.fx.experimental.optimization.optimize_for_inference changes OrderedDict outputs to immutable_dict in torch.export CallSpec
### 🐛 Describe the bug

When exporting a PyTorch module using `torch.export.export`, the `CallSpec's out_spec` correctly shows an `OrderedDict` for outputs that are `OrderedDict` in the model's forward. However, after applying `torch.fx.experimental.optimization.optimize_for_inference`, the same export replaces the `OrderedDict` in the output spec with an `immutable_dict`. This breaks downstream code that expects the output spec to match the model's output type (`OrderedDict`), and is inconsistent with the non-optimized export.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
class CustomModel(nn.Module):
    def __init__(self, kwargs=None):
        super(CustomModel, self).__init__()
        pass
    def forward(self, *args):
        x = args[0]
        import contextlib
        from collections import OrderedDict
        @contextlib.contextmanager
        def simple_context_manager():
          count = 0
          count += 1
          try:
            yield
          finally:
            count -= 1
        with simple_context_manager():
          nested_tuple = (F.relu(x), F.relu(x) + 1)
          nested_dict = OrderedDict([('out1', F.relu(x) + 2), ('out2', F.relu(x) + 3)])
          return (x, nested_tuple, nested_dict)

input_tensor = torch.randn(1,3,224)
model = CustomModel(input_tensor)
model(input_tensor)
model.eval()

# Export the module
ep = torch.export.export(model, (input_tensor,), strict=False)
print('CallSpec with optimization: \n',ep.call_spec)

# CallSpec(in_spec=TreeSpec(tuple, None, [TreeSpec(tuple, None, [*]),
#       TreeSpec(dict, [], [])]), out_spec=TreeSpec(tuple, None, [*,
#       TreeSpec(tuple, None, [*,
#         *]),
#       TreeSpec(OrderedDict, ['out1', 'out2'], [*,
#         *])]))

# See OrderedDict

# Now try with optimization
import torch.fx.experimental.optimization as optimization

optimization_config = {
    "conv_bn_fuse": True,
    "remove_dropout": True,
    "mkldnn_layout_optimize": False,
}
Optimzied_model = optimization.optimize_for_inference(model, optimization_config)
ep = torch.export.export(Optimzied_model, (input_tensor,), strict=False)

print('\n\nCallSpec with optimization: \n',ep.call_spec)

# CallSpec(in_spec=TreeSpec(tuple, None, [TreeSpec(tuple, None, [*]),
#       TreeSpec(dict, [], [])]), out_spec=TreeSpec(tuple, None, [*,
#       TreeSpec(tuple, None, [*,
#         *]),
#       TreeSpec(immutable_dict, ['out1', 'out2'], [*,
#         *])]))

# See immutable_dict
```

Please note that with `strict=True` the optimized spec changes from `immutable_dict` to `dict` which is still inconsistent with the expected `OrderedDict`

### Versions

Collecting environment information...
PyTorch version: 2.7.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 12 (bookworm) (x86_64)
GCC version: (Debian 12.2.0-14+deb12u1) 12.2.0
Clang version: Could not collect
CMake version: version 3.25.1
Libc version: glibc-2.36

Python version: 3.11.2 (main, Apr 28 2025, 14:11:48) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-6.1.0-38-amd64-x86_64-with-glibc2.36
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A16-4Q
Nvidia driver version: 570.133.20
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           45 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  10
On-line CPU(s) list:                     0-9
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 74F3 24-Core Processor
CPU family:                              25
Model:                                   1
Thread(s) per core:                      1
Core(s) per socket:                      5
Socket(s):                               2
Stepping:                                1
BogoMIPS:                                6387.99
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext invpcid_single ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero wbnoinvd arat umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor fsrm
Hypervisor vendor:                       VMware
Virtualization type:                     full
L1d cache:                               320 KiB (10 instances)
L1i cache:                               320 KiB (10 instances)
L2 cache:                                5 MiB (10 instances)
L3 cache:                                64 MiB (2 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-9
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; safe RET
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] numpy==2.0.2
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnx==1.18.0
[pip3] onnxscript==0.2.0
[pip3] torch==2.7.1
[pip3] torch-mlir==20250811.13
[pip3] torchaudio==2.7.1
[pip3] torchvision==0.22.1
[pip3] triton==3.3.1
[conda] Could not collect

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185150

FX TracerBase.create_arg previously handled OrderedDict through the generic dict path. Output nodes normalize dict aggregates to immutable_dict, so symbolic tracing a module that returns an OrderedDict produced a GraphModule that returned a plain dict. optimize_for_inference then exported that already-mutated output structure, causing torch.export CallSpec to report immutable_dict/dict instead of OrderedDict.

Fix this at the FX output boundary. When symbolic tracing finds an OrderedDict in an output, it flattens that output to its existing producer nodes and records the original pytree TreeSpec in an output-only code generator. Generated eager code reconstructs the original structure with pytree.tree_unflatten, preserving OrderedDict without adding an OrderedDict constructor node or any new graph-visible aggregate type.

TorchScript cannot represent OrderedDict or call pytree.tree_unflatten. For output structures whose keys are safely representable, the code generator therefore emits a torch.jit.is_scripting() branch with the equivalent dict/list/tuple literals. Eager execution and export continue to use the type-preserving pytree path. Unsupported and edge-case keys do not get a scripting branch, avoiding invalid generated source.

I considered special-casing export, but that would only hide the source bug because the GraphModule returned by FX tracing already had the wrong container type. I also considered carrying OrderedDict through a constructor node or a new FX aggregate, but that would expose a new non-Tensor operation or type to every downstream FX consumer. Keeping the type in output codegen metadata avoids both problems.

Fixes #161807
Generated by my agent

Test Plan:
- Targeted TestFX OrderedDict tracing tests loaded with torchvision intentionally blocked (passed)
- Direct repro covering symbolic_trace, optimize_for_inference, torch.jit.script, and torch.export.export with strict=False and strict=True (passed)
- python test/test_public_bindings.py TestPublicBindings.test_correct_module_names (passed)
- python -m py_compile torch/fx/_symbolic_trace.py torch/fx/graph.py torch/fx/graph_module.py test/test_fx.py (passed)
- lintrunner -a (passed)
- Full local TestVisionTracing invocation was blocked by the installed PyTorch/torchvision CUDA-version mismatch; the corresponding CI suite passed

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by zou3519 — CHANGES_REQUESTED** [review]
Agent: fx graph should not have intermediate nodes that return OrderedDict. I don't know if this PR is salvageable, if it isn't then please close it

**Review by jansel — COMMENTED** [review]
My agent says I addressed the review feedback by removing the intermediate call_function(collections.OrderedDict) reconstruction node. FX now carries OrderedDict outputs as an internal immutable ordered aggregate and reconstructs collections.OrderedDict only at the output boundary, including nested outputs under dicts; tests also cover that optimized graphs no longer contain an OrderedDict call_function node.

**Comment by jansel** [comment]
@claude review these changes

**Review by angelayi — APPROVED** [review]

**Comment by jansel** [comment]
My agent says I addressed the CI failures from test_public_bindings and TestFXAPIBackwardCompatibility by keeping the new OrderedDict aggregate helpers private and out of FX's backward-compatible public member set. I also added coverage for tuple-key OrderedDict outputs, including deepcopy and pickle reconstruction, and reran the targeted tests plus lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Review by zou3519 — CHANGES_REQUESTED** [review]
```
# IMPORTANT: Are you here because you are trying to proxy a new type into
        # the graph? Please Please Please contact someone on the PyTorch Compiler team;
        # the considerations are subtle.
        #
        # 1) When you add a new type, all of the downstream consumers and pass writers
        # need to handle the new type. torch.fx is intended to be easy to write
        # passes for, so we will push back against new types.
        # 2) In torch.compile's IR, there are only specific operations that go
        # into the graph. In particular, Tensor operations should go into the graph,
        # but non-Tensor operations shouldn't. What that means is that constructors
        # for new types *SHOULD NOT* become nodes in the FX graph.
```

**Review by jansel — COMMENTED** [review]
My agent says I addressed the OrderedDict review feedback by removing the internal FX aggregate approach. The updated PR now flattens OrderedDict-containing outputs to existing producer nodes and reconstructs the output structure with the existing pytree TreeSpec, so there is no OrderedDict constructor node and no new graph-visible aggregate type.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the new TestVisionTracing CI failures by making the OrderedDict output codegen scriptable for TorchScript-compatible dict outputs. Eager/export still reconstruct OrderedDict via the output TreeSpec, while TorchScript gets a guarded dict/list/tuple literal fallback only when the keys are representable safely; tuple/object/mixed/edge-case keys keep the existing pytree path. I reran the targeted FX repros, public bindings check, py_compile, and lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I added the two clarifying comments requested in the latest bot review: one documents why the TorchScript scripting branch emits plain dict literals, and the other explains the int64-min literal guard. The update is comment-only; I reran python -m py_compile torch/fx/graph.py and lintrunner -a.

**Comment by jansel** [comment]
My agent says I refreshed the PR description to match the current output-only pytree implementation, including the TorchScript fallback and current validation, instead of the obsolete constructor-node design.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
